core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java

```diff
@@ -378,4 +383,140 @@ public void run() {
+        for (Pair<RexNode, Object> value : Pair.zip(constExps, valueList)) {
+          reducedValues.add(
+              rexBuilder.makeLiteral(value.right, value.left.getType(), true));
+        }
+      } catch (RuntimeException e) {
+        // One or more of the expressions failed.
+        // Don't reduce any of the expressions.
+        reducedValues.addAll(constExps);
+      }
+    }
+    @Override public void setDataContext(DataContext dataContext) {
+      this.dataContext = dataContext;
+    }
+    @Override public Object[] execute() {
+      final RexNode exp = exps.get(0);
+      assert exp instanceof RexCall;
+      RexCall call = (RexCall) exp;
+      // NOTE: this isn't a correct substring function implementation, only for testing
+      if (!call.op.getName().equals("SUBSTRING")) {
+        throw new RuntimeException("not supported");
+      }
+      assert call.getOperands().get(0) instanceof RexInputRef;
+      assert call.getOperands().get(1) instanceof RexLiteral;
+      String str = (String) ((Object[]) dataContext.get("inputRecord"))[0];
+      Integer startIndex = ((RexLiteral) call.getOperands().get(1)).getValueAs(Integer.class);
+      Object[] values = new Object[1];
+      values[0] = str.substring(startIndex - 1);
+      return values;
+    }
+  }
+  /**
+   * User defined executor.
+   */
+  static class UserExecutor implements RexExecutor {
+    @Override public void reduce(RexBuilder rexBuilder, List<RexNode> constExps,
+        List<RexNode> reducedValues) {
+      this.getExecutable(rexBuilder, constExps, null).reduce(rexBuilder, constExps, reducedValues);
+    }
+    @Override public RexExecutable getExecutable(RexBuilder rexBuilder, List<RexNode> exps,
+        RelDataType rowType) {
+      return new UserExecutable(rexBuilder, exps);
+    }
+  }
+  @Test public void userExecutorTest() throws Exception {
+    check((rexBuilder, executor) -> {
+      Object[] values = new Object[1];
+      final DataContext testContext = new TestDataContext(values);
+      final RelDataTypeFactory typeFactory = rexBuilder.getTypeFactory();
+      final RelDataType varchar =
+          typeFactory.createSqlType(SqlTypeName.VARCHAR);
+      final RelDataType integer =
+          typeFactory.createSqlType(SqlTypeName.INTEGER);
+      // Calcite is internally creating the input ref via a RexRangeRef
+      // which eventually leads to a RexInputRef. So we are good.
+      final RexInputRef input = rexBuilder.makeInputRef(varchar, 0);
+      final RexNode lengthArg = rexBuilder.makeLiteral(3, integer, true);
+      final RexNode substr =
+          rexBuilder.makeCall(SqlStdOperatorTable.SUBSTRING, input,
+              lengthArg);
+      ImmutableList<RexNode> constExps = ImmutableList.of(substr);
+      final RelDataType rowType = typeFactory.builder()
+          .add("someStr", varchar)
+          .build();
+      RexExecutor useExecutor = new UserExecutor();
+      final RexExecutable exec =
```

Comment:
Hi @zhuwenzhuang ,could you add more test cases to cover your logic? I see only SUBSTRING case is added here.

Reply:
Hi @LakeShen. 
The userExecutorTest covered a user-defined RexExecutable/RexExecutor scenario. So, maybe adding more mock functions will not increase the code coverage.
Please let me know which part of code requires additional coverage with test cases. Thank you.